crates/core/tedge_agent/src/tedge_to_te_converter/actor.rs

```diff
@@ -0,0 +1,236 @@
+use clock::Timestamp;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json::Value;
+use std::collections::HashMap;
+use tedge_actors::Actor;
+use tedge_actors::MessageReceiver;
+use tedge_actors::RuntimeError;
+use tedge_actors::Sender;
+use tedge_actors::SimpleMessageBox;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct ThinEdgeAlarmData {
+    pub text: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[serde(with = "time::serde::rfc3339::option")]
+    pub time: Option<Timestamp>,
+
+    #[serde(default)]
+    pub severity: String,
+
+    #[serde(flatten)]
+    pub alarm_data: HashMap<String, Value>,
+}
+
+pub struct TedgetoTeConverterActor {
+    messages: SimpleMessageBox<MqttMessage, MqttMessage>,
+}
+
+#[async_trait]
+impl Actor for TedgetoTeConverterActor {
+    fn name(&self) -> &str {
+        "TedgetoTeConverterActor"
+    }
+
+    async fn run(&mut self) -> Result<(), RuntimeError> {
+        while let Some(message) = self.messages.recv().await {
+            {
+                self.process_mqtt_message(message).await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl TedgetoTeConverterActor {
+    pub fn new(messages: SimpleMessageBox<MqttMessage, MqttMessage>) -> Self {
+        Self { messages }
+    }
+
+    // Todo: Convert command topics
+    //tedge/commands/res/software/list	te/<identifier>/cmd/software_list/<cmd_id>
+    //tedge/commands/res/software/update	te/<identifier>/cmd/software_update/<cmd_id>
+    //tedge/+/commands/res/config_snapshot	te/<identifier>/cmd/config_snapshot/<cmd_id>
+    //tedge/+/commands/res/config_update	te/<identifier>/cmd/config_update/<cmd_id>
+    //tedge/+/commands/res/firmware_update	te/<identifier>/cmd/firmware_update/<cmd_id>
+    //tedge/commands/res/control/restart	te/<identifier>/cmd/restart/<cmd_id>
+    async fn process_mqtt_message(
+        &mut self,
+        message: MqttMessage,
+    ) -> Result<(), TedgetoTeConverterError> {
+        match message.topic.clone() {
+            topic if topic.name.starts_with("tedge/measurements") => {
+                self.convert_measurement(message).await?;
+            }
+            topic if topic.name.starts_with("tedge/events") => {
+                self.convert_event(message).await?;
+            }
+            topic if topic.name.starts_with("tedge/alarms") => {
+                self.convert_alarm(message).await?;
+            }
+            topic if topic.name.starts_with("tedge/health") => {
+                self.convert_health_status_message(message).await?;
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    // tedge/measurements -> te/device/main///m/
+    // tedge/measurements/child -> te/device/child///m/
+    async fn convert_measurement(
+        &mut self,
+        message: MqttMessage,
+    ) -> Result<(), TedgetoTeConverterError> {
```

Comment:
Changing the signature of this method would simplify the code a lot.
- keeping this function pure
- moving the effects (sending the message once converted) in one place (`run()`)
- removing the need for `TedgetoTeConverterError`.

```suggestion
    async fn convert_measurement(
        &mut self,
        message: MqttMessage,
    ) -> Option<MqttMessage> {
```

Reply:
In the case of `measurement, event, and health status` just need to update the `topic`, so it works fine. But in the case of `alarm` the `severity` has to be added to the `payload`. For this existing payload has to be deserialized using `serde`, which emits an error. Also, after adding the `severity` into the payload this has to be `serialized`, this also emits an error. So, I feel we need to return `Result<Message, Error>`.